File: scala/src/main/scala/arraymap/ArrayMap.scala
package arraymap

import scala.reflect.ClassTag
import scala.util.boundary

class ArrayMap[K: ClassTag, V: ClassTag] private(
                                                  private val keys: Array[Option[K]],
                                                  private val values: Array[Option[V]],
                                                  private var _currSize: Int
                                                ) {

  def this(capacity: Int) = this(new Array(capacity), new Array(capacity), 0)

  export keys.length as capacity

  def currSize: Int = _currSize

  def apply(k: K): Option[V] = {
    val idx = keys.indexOf(Some(k))
    Option.when(idx != -1) {
      values(idx)
    }.flatten
  }

  def update(k: K, v: V): Boolean = {
    val preSize = currSize
    val canAdd = preSize < capacity
    if (canAdd) {
      boundary {
        // ERROR: should be until instead of to (or better: for i <- keys.indices)
        for (i <- 0 to keys.length) {
          if (keys(i) == null) {
            keys(i) = Some(k)
            values(i) = Some(v)
          }
        }
      }
      _currSize = preSize + 1
    }
    canAdd
  }

  def remove(k: K): Boolean = {
    val preSize = currSize
    if (preSize > 0) {
      val idx = keys.indexOf(Some(k))
      if idx == -1 then false
      else {
        keys(idx) = None
        values(idx) = None
        _currSize = preSize - 1
        true
      }
    } else false
  }

}

File: java-index-checker/src/main/java/org/example/arraymap/ArrayMap.java
package org.example.arraymap;

import org.checkerframework.checker.index.qual.IndexOrHigh;
import org.checkerframework.checker.index.qual.NonNegative;
import org.checkerframework.checker.index.qual.SameLen;
import org.checkerframework.dataflow.qual.Pure;

public class ArrayMap<K, V> {
    private final K[] keys;
    private final V @SameLen("this.keys")[] values;
    private @IndexOrHigh("this.keys") int currSize;

    public ArrayMap(K[] keys, V @SameLen("#1")[] values) {
        this.keys = keys;
        this.values = values;
        this.currSize = 0;
    }

    public @Pure @NonNegative int capacity(){
        return this.keys.length;
    }

    public @IndexOrHigh("this.keys") int currentSize() {
        return this.currSize;
    }

    public V get(K k) {
        var idx = ArrayUtils.indexOf(keys, k);
        if (idx == -1) {
            return null;
        } else {
            return this.values[idx];
        }
    }

    public boolean put(K k, V v) {
        if (k == null || v == null) {
            throw new IllegalArgumentException();
        }
        var preSize = currentSize();
        if (preSize < this.keys.length) {
            // ERROR: should be < instead of <=
            for (int i = 0; i <= this.keys.length; i++) {
                if (this.keys[i] == null) {
                    this.keys[i] = k;
                    this.values[i] = v;
                    break;
                }
            }
            this.currSize = preSize + 1;
            return true;
        }
        return false;
    }

    public boolean remove(K k) {
        var preSize = currentSize();
        if (preSize > 0) {
            var idx = ArrayUtils.indexOf(this.keys, k);
            if (idx == -1) {
                return false;
            }
            this.keys[idx] = null;
            this.values[idx] = null;
            this.currSize = preSize - 1;
            return true;
        }
        return false;
    }

    @SuppressWarnings("unchecked")
    public static <K, V> ArrayMap<K, V> newEmptyOfCapacity(@NonNegative int capacity) {
        var keys = (K[]) new Object[capacity];
        var values = (V[]) new Object[keys.length];  // new Object[capacity] fails the VC on the next line
        return new ArrayMap<>(keys, values);
    }

}
